#164139686 Creating an article

kleva-j opened the pull request:
#### What does this PR do?
This PR implement the functionality of creating an article by a user.

#### Description of Task to be completed?
- Setup an endpoint to post articles.
- Ensures that all required inputs are sent.
- Provides swagger documentation for this task.

#### How should this be manually tested?
- Run `npm run start:dev` to start the application.
- Check .env.sample for the updated .env variables.
- Run `npm run start` to start the application.
- Make a POST request to `/api/v1/articles` with a JSON object containing the following fields: `title`, `description`, and `body`. They're all required. If one is not sent, there'll be an error.
- If this request is successful, you should receive a success message.

#### Any background context you want to provide?
- N/A

#### What are the relevant pivotal tracker stories?
#164139686

kleva-j committed:
feat(create_article): add functionality to create articles
- setup middlewares to validate article details and to authenticate user token
- create database models for articles
- setup routes for the endpoint to post an article
- write test for testing the functionality to create an article
- add documentation for this endpoint
[Delievers #164139686]

codinger41 commented on middleware/auth.js:
```diff
@@ -0,0 +1,22 @@
+  if (!authorization) {
+    return res.status(401).json({
+      success: false,
+      message: 'You are required to login to access this endpoint',
```
I think this error message should be something a user can understand. For example, `Unauthorized! You are required to be logged in to perform this operation.`. We have to assume users know nothing about endpoints.

kleva-j replied:
Totally understand, I will implement this immediately.

codinger41 commented on middleware/auth.js:
```diff
@@ -0,0 +1,22 @@
+    if (err) {
+      return res.status(401).json({
+        success: false,
+        message: 'Failed to authenticate',
```
I think we can also change this error message to something the user can understand also. For example, we can say `Your session has expired, please login again to continue`

kleva-j replied:
I get your point. That is a smart one.

olajide1234 commented on controllers/articles.js:
```diff
@@ -0,0 +1,30 @@
+        article: result
+      });
+    } catch (error) {
+      return res.status(500).json({ success: false, error });
```
Please return error in arrays e.g. `error: [error.message]`

kleva-j replied:
Yeah right.

codinger41 commented on middleware/verifyUser.js:
```diff
@@ -0,0 +1,12 @@
+import { User } from '../models';
+
+export default (req, res, next) => User.findById(req.user.id, { where: { id: req.user.id } })
```
What if we make this a class method for the `User` model? I learned recently that it's cool to have fat models and thin controllers, and that databases would run things faster. We can just add a `checkVerificationStatus` method that returns if the user is verified or not. Then in this middleware, we can have an async function that'll run the `checkVerificationSatatus` then return your error message or `next()`.

codinger41 commented on models/User.js:
```diff
@@ -47,7 +47,17 @@ module.exports = (sequelize, DataTypes) => {
+      associate(models) {
+        User.hasMany(models.Article, {
+          foreignKey: 'userId',
+          as: 'todoItems',
```
Little mistake here, shouldn't be `todoItems`.

kleva-j replied:
This was a mistake. Thanks for spotting it out.

olajide1234 commented on middleware/validation.js:
```diff
@@ -35,3 +35,21 @@ export const validateSignup = [
+    .exists()
+    .withMessage('Article should have a title.'),
+
+  check('description')
```
Consider making the description field optional.

kleva-j replied:
Form the rubric it states that the description is required input.

olajide1234 reviewed (commented):
Hi Mike, some comments for you. Thanks.

9jaswag commented on controllers/articles.js:
```diff
@@ -0,0 +1,30 @@
+        article: result
+      });
+    } catch (error) {
+      return res.status(409).json({ success: false, error: [error.message] });
```
@kleva-j why a 409 error here?

kleva-j replied:
I was a mistake, I have now fixed it.

9jaswag commented on models/User.js:
```diff
@@ -47,10 +47,19 @@ module.exports = (sequelize, DataTypes) => {
   User.hook('beforeValidate', (user) => {
     user.verificationId = shortId.generate();
-    user.password = bcrypt.hashSync(user.password, bcrypt.genSaltSync(10));
+    if (user.password) {
```
@kleva-j why do we have to check if a password was submitted?

codinger41 replied:
@9jaswag The `beforeValidate` hook runs for different methods like `User.update()`, and if it doesn't see the password field while trying to update the user's data, bcrypt will throw an error. So we decided to check if the password was supplied then hash it with bcrypt. If not, when a user wants to update his/her account, beforeValidate checks for password and if it's not provided, it'll not hash the password. So in this case, the password hashing happens only when a user signs up

9jaswag commented on routes/v1.js:
```diff
@@ -1,19 +1,21 @@
-  UserController.verifyAccount,
-);
+apiRoutes.route('/articles')
+  .post(Auth.verifyUser, isUserVerified, validateArticle, returnValidationErrors, createArticle);
```
@kleva-j there's a `validateArticle` middleware here. Wouldn't it be a good point to trim the article content, title, etc?

kleva-j replied:
Yes, I actually took care of trimming the article using a hook on the article model.

9jaswag replied:
@kleva-j Yeah I'm aware. I'm saying it'll be better you do it in the middleware, than in the db hook

9jaswag commented on package.json:
```diff
@@ -4,16 +4,15 @@
     "lint": "eslint routes models config test index.js",
-    "test": "npm run db:undo-migration && npm run db-migrate && mocha test/index.js --timeout=20000 --require babel-register --exit",
-    "start": "sequelize db:migrate && node dist/index",
+    "start": "sequelize db:migrate && npm run build && babel-node index.js",
```
@kleva-j why do we have `babel-node` back here?

kleva-j replied:
That's a mistake.
Fixed it.

9jaswag commented on package.json:
```diff
@@ -4,16 +4,15 @@
     "lint": "eslint routes models config test index.js",
-    "test": "npm run db:undo-migration && npm run db-migrate && mocha test/index.js --timeout=20000 --require babel-register --exit",
-    "start": "sequelize db:migrate && node dist/index",
+    "start": "sequelize db:migrate && npm run build && node index.js",
```
@kleva-j can you explain where the app is being served from based on this start script?

kleva-j replied:
I was a bit distracted when I was fixing the code. I was supposed to serve/start the app from the index file in the dist folder.

9jaswag commented on package.json:
```diff
@@ -4,16 +4,15 @@
     "lint": "eslint routes models config test index.js",
-    "test": "npm run db:undo-migration && npm run db-migrate && mocha test/index.js --timeout=20000 --require babel-register --exit",
-    "start": "sequelize db:migrate && node dist/index",
+    "start": "sequelize db:migrate && npm run build && node dist/ndex.js",
```
@kleva-j there's still an issue here. You're not referencing the same file. You should take time to go through all you've done to avoid anything you might have overlooked

9jaswag commented on package.json:
```diff
@@ -4,16 +4,15 @@
     "lint": "eslint routes models config test index.js",
-    "test": "npm run db:undo-migration && npm run db-migrate && mocha test/index.js --timeout=20000 --require babel-register --exit",
-    "start": "sequelize db:migrate && node dist/index",
+    "start": "sequelize db:migrate && npm run build && node dist/index.js",
```
@kleva-j should we be running the build script here?

kleva-j replied:
I wanted to get the new changes to the code to be transpiled when I run the app on development because babel-watch was not working well on my local machine.
I forgot to remove it for production.

9jaswag commented:
@kleva-j update the staging environment with the latest code in the PR.

kleva-j committed:
ft(add_image): enable image upload for article
- add middlewares to generate slug for articles
- create functions to validate article details
- update migrations file for articles
- create middleware to add and upload images
[Delivers #164139686]

codinger41 commented on controllers/articles.js:
```diff
@@ -0,0 +1,41 @@
+    const { id } = req.user;
+    const taglist = req.body.taglist || [];
+    try {
+      const ifUserExist = await User.findOne({ where: { id } });
```
@kleva-j  Let's change the name of this variable to something like `userExists`

codinger41 commented on helpers/trimArticle.js:
```diff
@@ -0,0 +1,7 @@
+  if (title) req.title = title.trim();
+  if (description) req.description = description.trim();
+  if (body) req.body = body.trim();
+  return next;
```
use `return next()`

codinger41 commented on middleware/addImage.js:
```diff
@@ -0,0 +1,28 @@
+
+export default async (req, res, next) => {
+  await imageParser(req, res, (error) => {
+    if (error) return res.json({ success: false, error });
```
The error here should be in our agreed format, an Array.

codinger41 commented on middleware/generateSlug.js:
```diff
@@ -0,0 +1,15 @@
+  const { title } = req.body;
+  const slug = generateSlug(req.body.title);
+  try {
+    const ifExist = await Article.findOne({ where: { slug } });
```
Let's change the name of this variable here to `articleExists`.

9jaswag commented on README.md:
```diff
@@ -1,6 +1,6 @@
 Authors Haven - A Social platform for the creative at heart.
 
-[![Build Status](https://travis-ci.com/andela/vidar-ah-backend.svg?branch=develop)](https://travis-ci.com/andela/vidar-ah-backend) [![Coverage Status](https://coveralls.io/repos/github/andela/vidar-ah-backend/badge.svg?branch=master)](https://coveralls.io/github/andela/vidar-ah-backend?branch=master) [![](https://img.shields.io/badge/Protected_by-Hound-a873d1.svg)](https://houndci.com)
+[![Build Status](https://travis-ci.com/andela/vidar-ah-backend.svg?branch=develop)](https://travis-ci.com/andela/vidar-ah-backend) [![Coverage Status](https://coveralls.io/repos/github/andela/vidar-ah-backend/badge.svg?branch=ft-create-user-article-%23164139686)](https://coveralls.io/github/andela/vidar-ah-backend?branch=ft-create-user-article-%23164139686) [![](https://img.shields.io/badge/Prote ...
```
@kleva-j  should this badge not reflect on the master/develop branch?

kleva-j replied:
No, I think it would when it gets merged. This is the coverage report for this branch.
